splink/duckdb/database_api.py

```diff
@@ -61,11 +61,7 @@ def _table_registration(self, input, table_name) -> None:
         elif isinstance(input, list):
             input = pd.DataFrame.from_records(input)
 
-        # Registration errors will automatically
-        # occur if an invalid data type is passed as an argument
-        self._execute_sql_against_backend(
-            f"CREATE TABLE {table_name} AS SELECT * FROM input"
-        )
+        self._con.register(table_name, input)
```

Comment:
This change has two effects:
1. It reduces memory usage, see script below
2. It means the table doesn't need to be cleaned up.  The `create table` version creates a materialised table in the database.  But since this is not a SplinkDataframe and is not registered in the intermediate table cache, it isn't cleaned up by `delete_tables_created_by_splink_from_db() ` (which iterates through  SplinkDataFrames that Splink knows about and if 'created_by_splink' is set to True, allows them to be deleted

<details>
<summary><details>
<summary>
Illustration of how register doesn't use memory but create table does
</summary>

```python
import os

import duckdb
import pandas as pd
import psutil


def get_size():
    process = psutil.Process(os.getpid())
    bytes = process.memory_info().rss  # in bytes
    factor = 1024
    for unit in ["", "K", "M", "G", "T", "P"]:
        if bytes < factor:
            return f"{bytes:.2f}{unit}B"
        bytes /= factor


print("Initial memory usage:", get_size())

df = pd.read_parquet(
    "/Users/robinlinacre/Documents/data_linking/ohid_docker/synthetic_1m.parquet"
)

print("Memory usage after loading DataFrame:", get_size())

con  = duckdb.connect()

con.register("registered_table", df)
print("Memory usage after registering DataFrame as a table in DuckDB:", get_size())

con.sql("CREATE TABLE my_table AS SELECT * FROM registered_table")

print("Memory usage after creating a new table from the registered table:", get_size())
```

</details>
Illustration of how register doesn't use memory but create table does
</summary>

```python
import os

import duckdb
import pandas as pd
import psutil


def get_size():
    process = psutil.Process(os.getpid())
    bytes = process.memory_info().rss  # in bytes
    factor = 1024
    for unit in ["", "K", "M", "G", "T", "P"]:
        if bytes < factor:
            return f"{bytes:.2f}{unit}B"
        bytes /= factor


print("Initial memory usage:", get_size())

df = pd.read_parquet(
    "/Users/robinlinacre/Documents/data_linking/ohid_docker/synthetic_1m.parquet"
)

print("Memory usage after loading DataFrame:", get_size())

con  = duckdb.connect()

con.register("registered_table", df)
print("Memory usage after registering DataFrame as a table in DuckDB:", get_size())

con.sql("CREATE TABLE my_table AS SELECT * FROM registered_table")

print("Memory usage after creating a new table from the registered table:", get_size())

Initial memory usage: 119.45MB
Memory usage after loading DataFrame: 865.04MB
Memory usage after registering DataFrame as a table in DuckDB: 867.31MB
Memory usage after creating a new table from the registered table: 1000.46MB
```

</details>

Reply:
In terms of what this does, see [here](https://duckdb.org/docs/api/python/data_ingestion#directly-accessing-dataframes-and-arrow-objects)

>  DuckDB also supports “registering” a DataFrame or Arrow object as a virtual table, comparable to a SQL VIEW. This is useful when querying a DataFrame/Arrow object that is stored in another way (as a class variable, or a value in a dictionary). Below is a Pandas example

Reply:
Yeah that makes sense. I originally switched to using tables when I implemented the `DatabaseAPI` as at that time it was difficult getting things working with some things being tables and some being views ([see note under 'DuckDB registration'](https://github.com/moj-analytical-services/splink/pull/1833)), and that was easiest for time being, but happy to go for the lighter approach if that works okay

Reply:
ah looks like there is still some sort of clash with views/tables. but yeah if that can be resolved i'm all in favour of the reduced-memory option